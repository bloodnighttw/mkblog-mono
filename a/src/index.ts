import path from "node:path";
import { type PluggableList, unified } from "unified";
import remark2rehype from "remark-rehype";
import remarkParse from "remark-parse";
import rehypeStringify from "rehype-stringify";
import fs from "node:fs";
import type { Root as HastRoot } from "hast";
import type { Root as MDRoot } from "mdast";
import frontmatter from "remark-frontmatter";
import { select } from "unist-util-select";
import yaml from "yaml";
import { z, type ZodRawShape } from "zod";

interface ContentOptions<T extends ZodRawShape> {
	root: string;
	matchRegex?: RegExp | "markdown";
	createSlug?: (from: string) => string;
	schema: T;

	mkBlogPlugins?: () => void;
	remarkPlugins?: PluggableList;
	rehypePlugins?: PluggableList;
}

interface YAMLNode {
	type: "yaml";
	value: string;
}

class Content<T> {
	readonly filepath: string;
	readonly slug: string;
	private readonly remarkPlugins: PluggableList;
	private readonly rehypePlugins: PluggableList;
	private readonly schema: z.ZodSchema<T>;

	constructor(
		filepath: string,
		slug: string,
		remarkPlugins: PluggableList,
		rehypePlugins: PluggableList,
		schema: z.Schema,
	) {
		this.filepath = filepath;
		this.slug = slug;
		this.remarkPlugins = remarkPlugins;
		this.rehypePlugins = rehypePlugins;
		this.schema = schema;
	}

	protected remarkProcessor() {
		return unified()
			.use(remarkParse)
			.use(frontmatter) // for metadata decode
			.use(this.remarkPlugins);
	}

	protected rehypeProcessor() {
		return unified().use(remark2rehype).use(this.rehypePlugins);
	}

	protected rehype2html() {
		return unified().use(rehypeStringify);
	}

	/**
	 * @protected to fetch the markdown and html ast, used internal.
	 */

	protected async mdast(): Promise<MDRoot> {
		const fileContent = await fs.promises.readFile(this.filepath, "utf-8");
		return this.remarkProcessor().parse(fileContent);
	}

	protected async hast(mdroot: MDRoot | null = null): Promise<HastRoot> {
		const temp = mdroot ?? (await this.mdast());
		return await this.rehypeProcessor().run(temp);
	}

	/**
	 *
	 * @param root the hast root, if it's null, we will fetch hast automatically
	 * @protected this method should be called internally
	 * @Return return raw html string of content.
	 */
	protected async html(root: HastRoot | null = null) {
		const temp = root ?? (await this.hast());
		return this.rehype2html().stringify(temp);
	}

	public async metadata(mdast: MDRoot | null = null): Promise<T> {
		const temp = mdast ?? (await this.mdast());
		const yamlContent = select("yaml", temp) as YAMLNode;

		if (yamlContent) {
			const metadata = yaml.parse(yamlContent.value);
			return this.schema.parse(metadata);
		}

		throw new Error("metadata not found");
	}
}

interface Collection<T> {
	posts: Content<T>[];
}


const markdownRegex = /\.md$/;

async function defineContent<T extends ZodRawShape>(
	options: ContentOptions<T>,
): Promise<Collection<T>> {
	const base = path.join(process.cwd(), options.root);
	const changeRegex =
		options.matchRegex === "markdown" || !options.matchRegex
			? markdownRegex
			: options.matchRegex;
	const files = (await fs.promises.readdir(base)).filter((file) =>
		changeRegex.test(file),
	);

	const posts = files.map((file) => {
		const filepath = path.join(base, file);
		const slug =
			options.createSlug?.(file) ?? file.replace(changeRegex, "");
		return new Content<T>(
			filepath,
			slug,
			options.remarkPlugins ?? [],
			options.rehypePlugins ?? [],
			z.object(options.schema),
		);
	});

	return { posts };
}
export type { ContentOptions, Collection };

export { defineContent, Content, z as zod };
