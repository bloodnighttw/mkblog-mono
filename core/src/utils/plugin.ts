import { PluggableList } from "unified";

interface MkblogPlugin {
	init?: () => void;
	remarkPlugins?: PluggableList;
	rehypePlugins?: PluggableList;
}

export { MkblogPlugin };
