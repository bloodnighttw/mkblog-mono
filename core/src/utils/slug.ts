import path from "node:path";

export function removeExtension(filepath: string) {
    return path.parse(filepath).name;
}

export function hashing(filepath: string) {
    throw new Error("Not implemented");
    return filepath;
}

export function mix(filepath: string) {
    throw new Error("Not implemented");
    return filepath;
}