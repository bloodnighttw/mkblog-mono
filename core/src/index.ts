import path from "node:path";
import {type PluggableList, unified} from "unified";
import remark2rehype from "remark-rehype";
import remarkParse from "remark-parse";
import rehypeStringify from "rehype-stringify";
import fs from "node:fs";
import type {Root as HastRoot} from "hast";
import type {Root as MDRoot, Yaml} from "mdast";
import frontmatter from "remark-frontmatter";
import {select} from "unist-util-select";
import yaml from "yaml";
import {z, type ZodRawShape} from "zod";
import fastGlob, {type Pattern} from "fast-glob";
import {hashing, mix, removeExtension} from "@/utils/slug";

interface ContentOptions<T extends ZodRawShape> {
    /**
     * @param include the glob pattern of the markdown files
     * @param createSlug the function to create the slug of the markdown file,
     * you can use "removeExtension" | "hashing" | "mix" | ((from: string) => string)
     *   - removeExtension: remove the extension of the file
     *   - hashing: hash the file path
     *   - mix: mix the file path and hash
     *   - ((from: string) => string): custom function to create the slug
     *   - default: removeExtension, if not provided
     * @param schema the schema of the metadata, using zod to define and validate the data.
     * @param mkBlogPlugins the function to create the plugins for the markdown blog, not implemented yet.
     * @param remarkPlugins the list of remark plugins
     * @param rehypePlugins the list of rehype plugins
     */
    include: Pattern | Pattern[]
	createSlug?: "removeExtension" | "hashing" | "mix" | ((from: string) => string);
	schema: T;

	mkBlogPlugins?: () => void;
	remarkPlugins?: PluggableList;
	rehypePlugins?: PluggableList;
}

/**
 * @Param T the type of the metadata, we use this to infer the type of the metadata.
 */
class Content<T> {
    /**
     * @param filepath the path of the markdown file
     * @param slug the slug of the markdown file, you can use this to generate the url
     * @param remarkPlugins the list of remark plugins
     * @param rehypePlugins the list of rehype plugins
     * @param schema the schema of the metadata, using zod to define and validate the data.
     * @constructor create a new instance of Content, you should use defineContent to create a collection of Content if
     * you don't know what this is doing.
     * @Return return a new instance of Content
     */
	readonly filepath: string;
	readonly slug: string;
	private readonly remarkPlugins: PluggableList;
	private readonly rehypePlugins: PluggableList;
	private readonly schema: z.ZodSchema<T>;

    constructor(
		filepath: string,
		slug: string,
		remarkPlugins: PluggableList,
		rehypePlugins: PluggableList,
		schema: z.Schema,
	) {
		this.filepath = filepath;
		this.slug = slug;
		this.remarkPlugins = remarkPlugins;
		this.rehypePlugins = rehypePlugins;
		this.schema = schema;
	}

    /**
     * @protected this method should be called internally, it's used to create a remark processor.
     * @Return return a unified processor which can be used to parse markdown content and apply remark plugins.
     */
	protected remarkProcessor() {
		return unified()
			.use(remarkParse)
			.use(frontmatter) // for metadata decode
			.use(this.remarkPlugins);
	}


    /**
     * @protected this method should be called internally, it's used to create a rehype processor.
     * @Return return a unified processor which can be used to convert mdast to hast and apply rehype plugins.
     */
	protected rehypeProcessor() {
		return unified().use(remark2rehype).use(this.rehypePlugins);
	}

    /**
     * @protected this method should be called internally, it's used to create a rehype to html processor.
     * @Return return a unified processor which can be used to convert hast to html string.
     */
	protected rehype2html() {
		return unified().use(rehypeStringify);
	}

    /**
     * @protected this method should be called internally, it's used to parse the markdown file and return the mdast.
     * @Return return the mdast of the markdown file.
     */
	protected async mdast(): Promise<MDRoot> {
		const fileContent = await fs.promises.readFile(this.filepath, "utf-8");
		return this.remarkProcessor().parse(fileContent);
	}


    /**
     * @protected this method should be called internally, it's used to convert the mdast to hast.
     * @param mdroot the mdast root, if it's null, we will fetch mdast automatically
     * @Return return the hast of the markdown file.
     */
	protected async hast(mdroot: MDRoot | null = null): Promise<HastRoot> {
		const temp = mdroot ?? (await this.mdast());
		return await this.rehypeProcessor().run(temp);
	}

	/**
	 * @param root the hast root, if it's null, we will fetch hast automatically
	 * @Return return raw html string of content.
	 */
	public async html(root: HastRoot | null = null) {
		const temp = root ?? (await this.hast());
		return this.rehype2html().stringify(temp);
	}

    /**
     * @param mdast the mdast root, if it's null, we will fetch mdast automatically
     * @Return return the metadata of the markdown file. we use zod to validate the metadata.
     * @throws if metadata not found or metadata is invalid.
     */
	public async metadata(mdast: MDRoot | null = null): Promise<T> {
		const temp = mdast ?? (await this.mdast());
		const yamlContent = select("yaml", temp) as Yaml | null;

		if (yamlContent) {
			const metadata = yaml.parse(yamlContent.value);
			return this.schema.parse(metadata);
		}

		throw new Error("metadata not found");
	}
}

interface Collection<T> {
	posts: Content<T>[];
}

async function defineContent<T extends ZodRawShape>(
	options: ContentOptions<T>,
): Promise<Collection<T>> {

    const files = await fastGlob(options.include);
    const slugFunc =
        options.createSlug === "removeExtension" ? removeExtension :
        options.createSlug === "hashing" ? hashing :
        options.createSlug === "mix" ? mix :
        options.createSlug ?? removeExtension;

	const posts = files.map((file) => {
		const filepath = path.join(process.cwd(), file);
		return new Content<T>(
			filepath,
			slugFunc(file),
			options.remarkPlugins ?? [],
			options.rehypePlugins ?? [],
			z.object(options.schema),
		);
	});

	return { posts };
}

export type { ContentOptions, Collection };

export { defineContent, Content, z as zod };
