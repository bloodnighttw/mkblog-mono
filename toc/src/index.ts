import {Content, MkblogPlugin} from "@mkblog/core";
import {Element, Root as HRoot} from "hast";
import {h} from "hastscript";
import rehypeSlug from "rehype-slug";
import {visit} from "unist-util-visit";

declare module "@mkblog/core" {
    interface Content<T> {
        rawTOC: () => Promise<HRoot>;
        toc: () => Promise<string>;
    }
}

const tocPlugin: MkblogPlugin = {
    init: () => {
        Content.prototype.rawTOC = async function() {
            const hast = await this.hast();
            const headings: { id: string, text: string, level: number }[] = [];

            visit(hast, 'element', (node: Element) => {
                if (/^h[1-6]$/.test(node.tagName)) {
                    const id = node.properties?.id as string ?? "";
                    const text = node.children
                        .filter(child => child.type === 'text')
                        .map(child => child.value)
                        .join();
                    const level = Number.parseInt(node.tagName[1], 10);
                    headings.push({ id, text, level });
                }
            });

            const buildTOC = (headings: { id: string, text: string, level: number }[], currentLevel = 1): Element => {
                const children: Element[] = [];
                while (headings.length > 0 && headings[0].level >= currentLevel) {

                    const heading = headings.shift()!;
                    const listItem: Element = h('li', { className: `toc-level-${heading.level}` }, [
                        h('a', { href: `#${heading.id}` }, heading.text)
                    ]);

                    if (headings.length > 0 && headings[0].level > heading.level) { // note: null or undefined < 1 === true
                        listItem.children.push(buildTOC(headings, heading.level + 1));
                    }
                    children.push(listItem);
                }
                return h('ul', { className: 'toc' }, children);
            };

            return h(null, buildTOC(headings));
        }

        Content.prototype.toc = async function() {
            const tocTree = await this.rawTOC();
            return this.rehype2html().stringify(tocTree);
        }
    },
    rehypePlugins: [rehypeSlug],
}

export default tocPlugin;