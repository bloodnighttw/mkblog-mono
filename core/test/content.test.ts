import * as path from "node:path";
import remarkGfm from "remark-gfm";
import { expect, test } from "vitest";
import { defineContent, zod } from "../src";
import { hashing, mix, removeExtension } from "../src/utils/slug";

/**
 * @param filepath the path of the markdown file
 * @Return the filename without extension
 */
const customSlugFunc = (filepath: string) => {
	return path.parse(filepath).name;
};

test("no yaml in markdown", async () => {
	const config = await defineContent({
		include: "./cases/unknown/*.md",
		createSlug: customSlugFunc,
		schema: {
			title: zod.string(),
		},
	});

	const post = config.posts.find((post) => post.slug === "no-yaml");
	expect(post).toBeDefined();

	await expect(config.posts[0].metadata()).rejects.toThrow(
		"metadata not found",
	);
});

test("valid metadata schema check", async () => {
	const config = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: customSlugFunc,
		schema: {
			title: zod.string(),
		},
	});

	const post = config.posts.find((post) => post.slug === "title");
	expect(post).toBeDefined();
	expect(await config.posts[0].metadata()).toBeDefined();
});

test("invalid metadata schema check", async () => {
	const config = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: customSlugFunc,
		schema: {
			title: zod.number(), // should be string
		},
	});

	const post = config.posts.find((post) => post.slug === "title");
	expect(post).toBeDefined();
	await expect(config.posts[0].metadata()).rejects.toThrow();
});

test("html check", async () => {
	const configWithGFM = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: customSlugFunc,
		schema: {
			title: zod.number(), // should be string
		},
		remarkPlugins: [remarkGfm],
	});

	const configWithoutGFM = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: customSlugFunc,
		schema: {
			title: zod.string(),
		},
	});

	const postWithGFM = configWithGFM.posts.find(
		(post) => post.slug === "title",
	);
	expect(postWithGFM).toBeDefined();
	const htmlWithGFM = await postWithGFM.html();
	expect(htmlWithGFM).toBeDefined();
	expect(htmlWithGFM).toMatchSnapshot();

	const postWithoutGFM = configWithoutGFM.posts.find(
		(post) => post.slug === "title",
	);
	expect(postWithoutGFM).toBeDefined();
	const htmlWithoutGFM = await postWithoutGFM.html();
	expect(htmlWithoutGFM).toBeDefined();
	expect(htmlWithoutGFM).toMatchSnapshot();

	expect(htmlWithGFM).not.toBe(htmlWithoutGFM); // the output should be different since we use remark-gfm
});

test("createSlug check", async () => {
	const testCase = "config.md";

	const removeExt = removeExtension(testCase);
	expect(removeExt).toBe("config");

	const hash = hashing(testCase);
	expect(hash).toBe("1xxt3xut3s4");

	const mixContent = `${removeExt}-${hash}`;
	expect(mix(testCase)).toBe(mixContent);
});

test("createSlug check with config", async () => {
	let config = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: "removeExtension",
		schema: {
			title: zod.number(), // should be string
		},
	});

	let post = config.posts.find((post) => post.slug === "title");
	expect(post).toBeDefined();

	config = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: "hashing",
		schema: {
			title: zod.number(), // should be string
		},
	});

	post = config.posts.find((post) => post.slug === "nudwg7rest");
	expect(post).toBeDefined();

	config = await defineContent({
		include: "./cases/a1/*.md",
		createSlug: "mix",
		schema: {
			title: zod.number(), // should be string
		},
	});

	post = config.posts.find((post) => post.slug === "title-nudwg7rest");
	expect(post).toBeDefined();
});

test("createSlug check with default", async () => {
	const config = await defineContent({
		include: "./cases/a1/*.md",
		schema: {
			title: zod.number(), // should be string
		},
	});

	const post = config.posts.find((post) => post.slug === "title");
	expect(post).toBeDefined();
	expect(post.slug).toBe("title");
});
