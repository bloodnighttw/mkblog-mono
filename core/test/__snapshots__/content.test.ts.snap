// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`html check 1`] = `
"<p>This is a test</p>
<h1>This is a header</h1>
<h2>This is a gfm test</h2>
<table>
<thead>
<tr>
<th>This is a test</th>
<th>This is a test</th>
</tr>
</thead>
<tbody>
<tr>
<td>This is a test</td>
<td>This is a test</td>
</tr>
</tbody>
</table>
<pre><code class="language-python">print("Hello World")
</code></pre>
<p><strong>Ricky Astley is never:</strong></p>
<ol>
<li>Never Gonna Give You Up</li>
<li>Never Gonna Let You Down</li>
<li>Never Gonna Run Around and Desert You</li>
</ol>
<p><code>And that's it.</code></p>
<p><a href="https://www.youtube.com/watch?v=dQw4w9WgXcQ">This is a link</a></p>"
`;

exports[`html check 2`] = `
"<p>This is a test</p>
<h1>This is a header</h1>
<h2>This is a gfm test</h2>
<p>| This is a test | This is a test |
|----------------|----------------|
| This is a test | This is a test |</p>
<pre><code class="language-python">print("Hello World")
</code></pre>
<p><strong>Ricky Astley is never:</strong></p>
<ol>
<li>Never Gonna Give You Up</li>
<li>Never Gonna Let You Down</li>
<li>Never Gonna Run Around and Desert You</li>
</ol>
<p><code>And that's it.</code></p>
<p><a href="https://www.youtube.com/watch?v=dQw4w9WgXcQ">This is a link</a></p>"
`;
